crates/biome_service/src/settings.rs

```diff
@@ -589,36 +590,159 @@ pub struct FilesSettings {
+            (None, None) => return None,
+        })
+    }
+
+    /// Checks where if the current file is a recognised file for the current VCS client
+    pub fn is_ignore_file(&self, path: &Utf8Path) -> bool {
+        path.file_name().is_some_and(|file_name| {
+            self.client_kind
+                .is_some_and(|client_kind| client_kind.ignore_files().contains(&file_name))
+        })
+    }
+
+    /// Stores the contents found in the ignore file.
+    pub fn store_ignore_patterns(
+        &mut self,
+        path: &Utf8Path,
+        patterns: &[&str],
+    ) -> Result<(), WorkspaceError> {
+        match self.client_kind {
+            Some(VcsClientKind::Git) => {
+                let git_ignore = VcsIgnoredMatches::git_ignore(path, patterns)?;
+                if let Some(ignore_matches) = self.ignore_matches.as_mut() {
+                    ignore_matches.insert_git_match(git_ignore);
+                } else {
+                    self.ignore_matches = Some(VcsIgnoredMatches::Git {
+                        root: git_ignore,
+                        nested: vec![],
+                    });
+                }
+            }
+            None => {}
+        };
+
+        Ok(())
+    }
+}
+
+#[derive(Clone, Debug)]
+pub enum VcsIgnoredMatches {
+    Git {
+        // Represents the `.gitignore` file at the root of the project
+        root: Gitignore,
+        // The list of nested `.gitignore` files found inside the project
+        nested: Vec<Gitignore>,
+    },
+}
+
+impl VcsIgnoredMatches {
+    #[instrument(level = "debug", skip(self, path, is_dir), fields(result))]
+    pub fn is_ignored(&self, path: &Utf8Path, is_dir: bool) -> bool {
+        match self {
+            VcsIgnoredMatches::Git { root, nested } => {
+                root.matched(path, is_dir).is_ignore()
+                    || nested.iter().any(|gitignore| {
+                        let ignore_directory = if gitignore.path().is_file() {
+                            // SAFETY: if it's a file, it always has a parent
+                            gitignore.path().parent().unwrap()
+                        } else {
+                            gitignore.path()
+                        };
+                        let result = if let Ok(stripped_path) = path.strip_prefix(ignore_directory)
+                        {
+                            gitignore.matched(stripped_path, is_dir).is_ignore()
+                        } else {
+                            false
+                        };
+                        tracing::Span::current().record("result", result);
+                        result
+                    })
+            }
+        }
+    }
+
+    pub fn insert_git_match(&mut self, git_ignore: Gitignore) {
+        match self {
+            VcsIgnoredMatches::Git { nested, .. } => {
+                nested.push(git_ignore);
+            }
+        }
+    }
+
+    fn git_ignore(path: &Utf8Path, matches: &[&str]) -> Result<Gitignore, WorkspaceError> {
```

Comment:
Nit: this can also be "patterns".
```suggestion
    fn git_ignore(path: &Utf8Path, patterns: &[&str]) -> Result<Gitignore, WorkspaceError> {
```